psi4/src/psi4/dlpno/mp2.h

```diff
@@ -0,0 +1,180 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Psi4: an open-source quantum chemistry software package
+ *
+ * Copyright (c) 2007-2021 The Psi4 Developers.
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#ifndef PSI4_SRC_DLPNO_MP2_H_
+#define PSI4_SRC_DLPNO_MP2_H_
+
+#include "sparse.h"
+
+#include "psi4/libmints/wavefunction.h"
+
+namespace psi {
+namespace dlpno {
+
+// Equations refer to Pinski et al. (JCP 143, 034108, 2015; DOI: 10.1063/1.4926879)
+
+class DLPNOMP2 : public Wavefunction {
+   protected:
+
+    /// threshold for PAO domain size
+    double T_CUT_DO;
+
+    /// threshold for PNO truncation
+    double T_CUT_PNO;
+
+    /// auxiliary basis
+    std::shared_ptr<BasisSet> ribasis_;
+    SharedMatrix full_metric;
+
+    /// localized molecular orbitals (LMOs)
+    SharedMatrix C_lmo;
+    SharedMatrix F_lmo;
+
+    /// projected atomic orbitals (PAOs)
+    SharedMatrix C_pao;
+    SharedMatrix F_pao;
+    SharedMatrix S_pao;
+
+    /// differential overlap integrals (EQ 4)
+    SharedMatrix DOI_ij; // LMO/LMO
+    SharedMatrix DOI_iu; // LMO/PAO
+
+    // approximate LMO/LMO pair energies from dipole integrals (EQ 17)
+    SharedMatrix e_actual; ///< actual approximate pair energy
+    SharedMatrix e_linear; ///< upper bound to approximate pair energies
+
+    /// LMO/PAO three-index integrals
+    std::vector<SharedMatrix> qia;
+
+    /// pair natural orbitals (PNOs)
+    std::vector<SharedMatrix> K_iajb;  ///< exchange operators (i.e. (ia|jb) integrals)
+    std::vector<SharedMatrix> T_iajb;  ///< amplitudes
+    std::vector<SharedMatrix> Tt_iajb; ///< antisymmetrized amplitudes
+    std::vector<SharedMatrix> R_iajb;  ///< LMP2 residuals
```

Comment:
While I'm open to second opinions on this, I don't understand why `R_iajb` needs to be a member variable. It makes more sense to me to create the variable during `lmp2_iterations` and pass it by reference to the norm-measuring function. There's no need for this variable to exist outside of `lmp2_iterations`.

By contrast, I can get why exchange integrals or amplitudes would be wavefunction variables.

Reply:
Good point, the residuals should be effectively zero by the end of the iterations, and therefore aren't very useful to keep around. I removed them from the wavefunction.